### ISSUE TITLE
RS0030 Banned namespace symbol not reported

### ISSUE THREAD

**Issue by wachulski** [dotnet/roslyn-analyzers#2748]
#### Analyzer package

Example: [Microsoft.CodeAnalysis.BannedApiAnalyzers](https://www.nuget.org/packages/Microsoft.CodeAnalysis.BannedApiAnalyzers)

#### Package Version

Example: v2.9.4 (Latest)

#### Diagnostic ID
RS0030

#### Repro steps

BannedSymbols.txt: 
```
N:BN
```
.cs:
```csharp
namespace BN
{
    public class BC { }
}

namespace N
{
    public class C
    {
        BN.BC BC => new BN.BC();
    }
}
```

#### Expected behavior

Both occurrences of `BC` should be reported as contained within symbol `BN` which is banned.

#### Actual behavior

No diagnostics reported.

**Comment by genlu** [comment]
Using namespace as an banned symbol is currently not supported. Could you please provide a little more detail on your use case? In what scenario this is preferred than listing all contained types as banned symbols?

Also, in terms of behavior of banning namespace, do you want to ban the use of any types contained in the namespace, or also declaring such namespace in source?

**Comment by wachulski** [comment]
It could help architectural constraints maintenance by disallowing certain namespaces in the project. ATM some people write reflexive code in tests to assert on certain assembly reference connections. E.g. `Acme.Domain` should not use anything from `NHibernate` nor `Microsoft.EntityFrameworkCore` nor `Acme.Infrastructure`. There are some tools on the market like https://www.ndepend.com/docs/cqlinq-features#Dependencies or Mono.Cecil but with the native support of the compiler it'll become simpler.

Rather the former so to prevent external namespaces/types from creeping into the assembly.

Since namespaces are hierarchical, so banning `Acme.Infrastructure` should prevent all of the form: `Acme.Infrastructure.DataAccess`, `Acme.Infrastructure.RemoteServices` etc. Additionally with this approach we could ban multiple nugets/libs at once providing their names start with the same namespace. IMO effective. Is the potential impl complex?

**Comment by samtrion** [comment]
This is certainly also very helpful in the area of performance optimizations.

In various projects we fight for every `ns` in the code execution, but here are often standard implementations, which make our lives much more difficult, such as `System.Linq`.
Often this namespace is used without regard to the performance penalty associated with it. However, in a truly performance critical project, it is necessary to dispense with the namespace and its associated methods.
Based on team efforts, this is certainly feasible, but with a change in team personnel, the effort begins again.

**Comment by nxrighthere** [comment]
We would like to have such a feature to ban any types contained in namespaces such as `System.Threading` in cases where user code is submitted to a server, compiled, and executed from there remotely.

**Comment by nxrighthere** [comment]
@genlu Could you please consider implementing this feature?

**Comment by Bouke** [comment]
I have created pull request #6521 to implement this missing feature.

**Comment by jkoritzinsky** [comment]
Fixed with #6521
